### PR TITLE
Refactored Employee Type service and controller

### PR DESCRIPTION
**Pull request by ReitumetseMoloto**
Coverage for Service
image

Coverage for Controller
image

### CODE REVIEW

File: HRIS.Services.Tests/Services/EmployeeTypesServiceUnitTest.cs [hunk 1]
```diff
@@ -16,56 +16,79 @@ public class EmployeeTypesServiceUnitTest
     private readonly Mock<IUnitOfWork> _dbMock;
     private readonly Mock<IEmployeeTypeService> _employeeTypeServiceMock;
     private readonly EmployeeTypeService employeeTypeService;
```
Comment (EduanO) [inline HRIS.Services.Tests/Services/EmployeeTypesServiceUnitTest.cs]:
```suggestion
    private readonly EmployeeTypeService _employeeTypeService;
```

File: HRIS.Services.Tests/Services/EmployeeTypesServiceUnitTest.cs [hunk 2]
```diff
@@ -16,56 +16,79 @@ public class EmployeeTypesServiceUnitTest
...
+    private readonly Mock<AuthorizeIdentityMock> _identity;
 
     public EmployeeTypesServiceUnitTest()
     {
+        _identity = new Mock<AuthorizeIdentityMock>();
         _dbMock = new Mock<IUnitOfWork>();
         _employeeTypeServiceMock = new Mock<IEmployeeTypeService>();
-        employeeTypeService = new EmployeeTypeService(_dbMock.Object);
+        employeeTypeService = new EmployeeTypeService(_dbMock.Object, _identity.Object);
```
Comment (EduanO) [inline HRIS.Services.Tests/Services/EmployeeTypesServiceUnitTest.cs]:
```suggestion
        _employeeTypeService = new EmployeeTypeService(_dbMock.Object, _identity.Object);
```

File: HRIS.Services.Tests/Services/EmployeeTypesServiceUnitTest.cs [hunk 3]
```diff
@@ -125,21 +148,42 @@ public async Task DeleteEmployeeTypeTestSuccess()
...
+        _identity.Setup(x => x.EmployeeId).Returns(1);
 
-        _dbMock.Setup(e => e.EmployeeType.Get(It.IsAny<Expression<Func<EmployeeType, bool>>>()))
-               .Returns(employeeTypeList.ToMockIQueryable());
+        _dbMock.Setup(x => x.EmployeeType.Any(It.IsAny<Expression<Func<EmployeeType, bool>>>()))
+               .ReturnsAsync(true);
         _dbMock.Setup(x => x.EmployeeType.Delete(employeeTypeList[0].Id)).ReturnsAsync(employeeType);
 
-        var result = await employeeTypeService.DeleteEmployeeType(employeeTypeList[0].Name!);
+        var result = await employeeTypeService.DeleteEmployeeType(employeeTypeList[0].Id!);
```
Comment (EduanO) [inline HRIS.Services.Tests/Services/EmployeeTypesServiceUnitTest.cs]:
```suggestion
        var result = await employeeTypeService.DeleteEmployeeType(employeeTypeList[0].Id);
```

File: HRIS.Services.Tests/Services/EmployeeTypesServiceUnitTest.cs [hunk 4]
```diff
@@ -125,21 +148,42 @@ public async Task DeleteEmployeeTypeTestSuccess()
...
+        };
+
+        _dbMock.Setup(x => x.EmployeeType.Any(It.IsAny<Expression<Func<EmployeeType, bool>>>()))
+               .ReturnsAsync(false);
+
+        _dbMock.Setup(r => r.EmployeeType.Delete(employeeTypeList[0].Id)).ReturnsAsync(employeeType);
+
+        await Assert.ThrowsAsync<CustomException>(() => employeeTypeService.DeleteEmployeeType(employeeTypeDto.Id));
+
+        _dbMock.Verify(x => x.EmployeeType.Delete(employeeTypeList[0].Id!), Times.Never);
```
Comment (EduanO) [inline HRIS.Services.Tests/Services/EmployeeTypesServiceUnitTest.cs]:
```suggestion
        _dbMock.Verify(x => x.EmployeeType.Delete(employeeTypeList[0].Id), Times.Never);
```

File: HRIS.Services/Interfaces/IEmployeeTypeService.cs [hunk 5]
```diff
@@ -9,19 +9,19 @@ public interface IEmployeeTypeService
...
-    Task<EmployeeTypeDto> SaveEmployeeType(EmployeeTypeDto employeeTypeDto);
+    Task<EmployeeTypeDto> CreateEmployeeType(EmployeeTypeDto employeeTypeDto);
 
     /// <summary>
     ///     Delete Employee Type
     /// </summary>
     /// <param name="name"></param>
     /// <returns></returns>
-    Task<EmployeeTypeDto> DeleteEmployeeType(string name);
+    Task<EmployeeTypeDto> DeleteEmployeeType(int Id);
```
Comment (EduanO) [inline HRIS.Services/Interfaces/IEmployeeTypeService.cs]:
```suggestion
    Task<EmployeeTypeDto> DeleteEmployeeType(int id);
```

File: HRIS.Services/Interfaces/IEmployeeTypeService.cs [hunk 6]
```diff
@@ -9,19 +9,19 @@ public interface IEmployeeTypeService
...
 
     /// <summary>
     ///     Get Employee Type
     /// </summary>
-    /// <param name="name"></param>
+    /// <param Id="Id"></param>
     /// <returns></returns>
     Task<EmployeeTypeDto> GetEmployeeType(string name);
```
Comment (EduanO) [inline HRIS.Services/Interfaces/IEmployeeTypeService.cs]:
```suggestion
    Task<EmployeeTypeDto> GetEmployeeTypeByName(string name);
```

File: HRIS.Services/Interfaces/IEmployeeTypeService.cs [hunk 7]
```diff
@@ -9,19 +9,19 @@ public interface IEmployeeTypeService
...
     /// <param name="employeeTypeDto"></param>
     /// <returns></returns>
-    Task<EmployeeTypeDto> SaveEmployeeType(EmployeeTypeDto employeeTypeDto);
+    Task<EmployeeTypeDto> CreateEmployeeType(EmployeeTypeDto employeeTypeDto);
 
     /// <summary>
     ///     Delete Employee Type
     /// </summary>
     /// <param name="name"></param>
     /// <returns></returns>
```
Comment (EduanO) [inline HRIS.Services/Interfaces/IEmployeeTypeService.cs]:
```suggestion
        /// <summary>
        ///     Deletes an existing employee type by id.
        /// </summary>
        /// <param name="id">The id of the employee type to delete.</param>
        /// <returns>The deleted EmployeeTypeDto object.</returns>
```

Just requires some updates on summary info for all

File: HRIS.Services/Services/EmployeeTypeService.cs [hunk 8]
```diff
@@ -1,5 +1,7 @@
-﻿using HRIS.Models;
+﻿using Auth0.ManagementApi.Models;
```
Comment (EduanO) [inline HRIS.Services/Services/EmployeeTypeService.cs]:
```suggestion
```

File: HRIS.Services/Services/EmployeeTypeService.cs [hunk 9]
```diff
@@ -9,27 +11,48 @@ namespace HRIS.Services.Services;
...
 
-    public EmployeeTypeService(IUnitOfWork db)
+    public EmployeeTypeService(IUnitOfWork db, AuthorizeIdentity identity)
     {
         _db = db;
+        _identity = identity;
     }
 
-    public async Task<EmployeeTypeDto> SaveEmployeeType(EmployeeTypeDto employeeTypeDto)
+    public async Task<bool> EmployeeTypeExists(int Id)
```
Comment (EduanO) [inline HRIS.Services/Services/EmployeeTypeService.cs]:
```suggestion
    public async Task<bool> EmployeeTypeExists(int id)
```

File: HRIS.Services/Services/EmployeeTypeService.cs [hunk 10]
```diff
@@ -9,27 +11,48 @@ namespace HRIS.Services.Services;
...
+            throw new CustomException("Unauthorized Access.");
+
         var newEmployeeType = await _db.EmployeeType
                                        .Add(new EmployeeType(employeeTypeDto));
 
         return newEmployeeType.ToDto();
     }
 
-    public async Task<EmployeeTypeDto> DeleteEmployeeType(string name)
+    public async Task<EmployeeTypeDto> DeleteEmployeeType(int Id)
```
Comment (EduanO) [inline HRIS.Services/Services/EmployeeTypeService.cs]:
```suggestion
    public async Task<EmployeeTypeDto> DeleteEmployeeType(int id)
```

**Review by EduanO — CHANGES_REQUESTED** [review]

**Review by MathildaShongwe15 — APPROVED** [review]

File: HRIS.Services/Interfaces/IEmployeeTypeService.cs (hunk 5 above)
Comment (EduanO) [reply HRIS.Services/Interfaces/IEmployeeTypeService.cs]: Parameters are always Camel case so it needs to be `id` not `Id`

File: HRIS.Services/Interfaces/IEmployeeTypeService.cs (hunk 6 above)
Comment (EduanO) [reply HRIS.Services/Interfaces/IEmployeeTypeService.cs]: The method needs to be more description on gets.

`GetEmployeeType` does not describe how to get employ, hence something like `GetEmployeeTypeByName` fits the actions of the method better.

File: HRIS.Services/Interfaces/IEmployeeTypeService.cs [hunk 11]
```diff
@@ -5,24 +5,24 @@ namespace HRIS.Services.Interfaces;
...
+    /// <returns>The deleted EmployeeTypeDto object.</returns>
+    Task<EmployeeTypeDto> DeleteEmployeeType(int Id);
 
     /// <summary>
-    ///     Get Employee Type
+    ///     Get Employee Type by Name
     /// </summary>
-    /// <param name="name"></param>
-    /// <returns></returns>
+    /// <param Id="Id">The id of the employe type to get</param>
```
Comment (EduanO) [inline HRIS.Services/Interfaces/IEmployeeTypeService.cs]:
```suggestion
    /// <param Id="name">The name of the employee type to get</param>
```

**Review by EduanO — CHANGES_REQUESTED** [review]

File: HRIS.Services.Tests/Services/EmployeeTypesServiceUnitTest.cs [hunk 12]
```diff
@@ -15,57 +15,80 @@ public class EmployeeTypesServiceUnitTest
...
+
+        _dbMock.Verify(x => x.EmployeeType.Add(It.IsAny<EmployeeType>()), Times.Never);
+        _dbMock.Verify(x => x.EmployeeType.FirstOrDefault(It.IsAny<Expression<Func<EmployeeType, bool>>>()), Times.Never);
+    }
+
+    [Fact]
+    public async Task CreateEmployeeTypeTestUnauthorized()
+    {
+        var employeeType = EmployeeTypeTestData.DeveloperType;
+        var employeeTypeDto = EmployeeTypeTestData.DeveloperType.ToDto();
```
Comment (bseptember) [inline HRIS.Services.Tests/Services/EmployeeTypesServiceUnitTest.cs]: Refactor: We don't need to repeat these two lines of code in all the unit tests. We can set it once in the constructor and then use the variables where needed. Double check all the unit tests do not repeat code.

For example:
 
 public EmployeeTypesServiceUnitTest()
    {
    ...
    _employeeType = EmployeeTypeTestData.DeveloperType;
    _emloyeeTypeDto = EmployeeTypeTestData.DeveloperType.ToDto();
    _employeeTypes = new List<EmployeeType>
        {
           new() { Id = 1, Name = "CAM" },
           new() { Id = 2, Name = "Designer" },
           new() { Id = 3, Name = "Developer" }
        };
    }

**Review by bseptember — CHANGES_REQUESTED** [review]

**Review by EduanO — APPROVED** [review]

File: HRIS.Services.Tests/Services/EmployeeTypesServiceUnitTest.cs (hunk 12 above)
Comment (EduanO) [reply HRIS.Services.Tests/Services/EmployeeTypesServiceUnitTest.cs]: Agree

**Review by bseptember — APPROVED** [review]

**Pull request merged**